streaming/base/storage/download.py

```diff
@@ -35,25 +37,202 @@
+        """Get the downloader for the remote path.
 
-    def _download_file(unsigned: bool = False,
-                       extra_args: Optional[dict[str, Any]] = None) -> None:
-        """Download the file from AWS S3 bucket. The bucket can be either public or private.
+        Args:
+            remote (str | None): Remote path.
+
+        Returns:
+            CloudDownloader: Downloader for the remote path.
+
+        Raises:
+            ValueError: If the remote path is not supported.
+        """
+        if remote_dir is None:
+            return _DEFAULT_DOWNLOADER()
+
+        prefix = urllib.parse.urlparse(remote_dir).scheme
+        if prefix == 'dbfs' and remote_dir.startswith('dbfs:/Volumes'):
+            prefix = 'dbfs-uc'
+
+        if prefix not in DOWNLOADER_MAPPINGS:
+            raise ValueError(f'Unsupported remote path: {remote_dir}')
+
+        return DOWNLOADER_MAPPINGS[prefix]()
+
+    @classmethod
+    def direct_download(cls, remote: Optional[str], local: str, timeout: float = 60.0) -> None:
+        """Directly download a file from remote storage to local filesystem.
 
         Args:
-            unsigned (bool, optional):  Set to True if it is a public bucket.
-                Defaults to ``False``.
-            extra_args (Dict[str, Any], optional): Extra arguments supported by boto3.
-                Defaults to ``None``.
+            remote (str | None): Remote path.
+            local (str): Local path.
+            timeout (float): How long to wait for file to download before raising an exception.
+
+        Raises:
+            ValueError: If the remote path is not provided while local does not exist or remote
+                path is not supported.
         """
+        downloader = cls.get(remote)
+        downloader.download(remote, local, timeout)
+        downloader.clean_up()
+
+    def download(self, remote: Optional[str], local: str, timeout: float = 60.0) -> None:
+        """Download a file from remote storage to local filesystem.
+
+        Args:
+            remote (str | None): Remote path.
+            local (str): Local path.
+            timeout (float): How long to wait for file to download before raising an exception.
+
+        Raises:
+            ValueError: If the remote path does not contain the expected prefix or remote is
+                not provided while local does not exist.
+        """
+        if os.path.exists(local):
+            return
+
+        if not remote:
+            raise ValueError(
+                'In the absence of local dataset, path to remote dataset must be provided')
+
+        if sys.platform == 'win32':
+            remote = pathlib.PureWindowsPath(remote).as_posix()
+            local = pathlib.PureWindowsPath(local).as_posix()
+
+        local_dir = os.path.dirname(local)
+        os.makedirs(local_dir, exist_ok=True)
+
+        self._validate_remote_path(remote)
+        self._download_impl(remote, local, timeout)
```

Comment:
Should we just say `_download_file`?

Reply:
Hmm, how about `_download_file_impl`. The reason being is that I think having the word `impl` as part of an abstract/interface communicates to the developer that this part is to be implemented. However you're right in the case that having `file` is more reflective of what it is doing.

Reply:
i like `_download_file_impl` as well